shell/platform/darwin/macos/framework/Source/FlutterView.mm

```diff
@@ -20,14 +33,73 @@ - (instancetype)initWithFrame:(NSRect)frame
   if (self) {
     self.openGLContext = [[NSOpenGLContext alloc] initWithFormat:shareContext.pixelFormat
                                                     shareContext:shareContext];
+
+    [self setWantsLayer:YES];
+
+    // Layer for content (which will be set by surfaceManager). This is separate from
+    // self.layer because it needs to be flipped vertically (using layer.sublayerTransform)
+    contentLayer = [[CALayer alloc] init];
+    [self.layer addSublayer:contentLayer];
+
+    resizeSynchronizer = [[FlutterResizeSynchronizer alloc] initWithDelegate:self];
+    surfaceManager = [[FlutterSurfaceManager alloc] initWithLayer:contentLayer
+                                                    openGLContext:self.openGLContext];
+
     _reshapeListener = reshapeListener;
-    self.wantsBestResolutionOpenGLSurface = YES;
   }
   return self;
 }
 
+- (void)resizeSynchronizerCommit:(FlutterResizeSynchronizer*)synchronizer {
+  [self.openGLContext makeCurrentContext];
+  glFlush();
+  [NSOpenGLContext clearCurrentContext];
+
+  [CATransaction begin];
+  [CATransaction setDisableActions:YES];
+  self.layer.frame = self.bounds;
+  self.layer.sublayerTransform = CATransform3DTranslate(CATransform3DMakeScale(1, -1, 1), 0,
+                                                        -self.layer.bounds.size.height, 0);
+  contentLayer.frame = self.layer.bounds;
+
+  [surfaceManager swapBuffers];
+
+  [CATransaction commit];
+}
+
+- (int)getFrameBufferIdForSize:(CGSize)size {
+  if ([resizeSynchronizer shouldEnsureSurfaceForSize:size]) {
```

Comment:
This reads like the following:

1. If resizing is in progress:
   a. If the size of the backing store matches the requested frame size, return true, and we end up making sure a surface of the desired size exists.
   b. **If it doesn't match**, we still end up returning the framebuffer id, this is fine because we know this layer tree will be discarded by the discard callback.
2. If resizing is not in progress `shouldEnsureSurfaceForSize:size` will return `false` and we just return the fbo id.

Is that right? It might be worth mentioning about the discard callback here as it isn't very obvious.

Reply:
Well, discard callback helps reduce the number of instances where we rasterize wrong size tree, but it can't eliminate those that were already scheduled on raster thread. So those will still be painted, to the original size surface. It doesn't matter that the surface size doesn't match, because for these we never call glFlush and/or swap buffers. That's handled by `[FlutterResizeSynchronizer requestCommit]`.

Reply:
1. Why shouldn't we always call `surfaceManager.ensureSurfaceSize`?
2. What happens if we move this call to `resizeSynchronizerCommit`?

Reply:
1. There's no reason to recreate surfaces with wrong size, because whatever we paint to those surfaces will be discarded anyway. It's just additional overhead.

2. Commit is called after painting is done. You want to have the surface available before that (when you request the FBO). Also, all commit requests are ignored until `[ResizeSynchronizer shouldEnsureSurfaceForSize:]` is called with the right size. Otherwise you get resizing artifacts.